### PR TITLE
Add coverage build and upload unit-test coverage to codecov.io

### LINKED ISSUE #120: Add test coverage to CI.
I’m nearly there with `gcov` on the linux build → [codecov.io](https://codecov.io) (some [other UCL projects use this](https://app.codecov.io/gh/ucl)).
➕  We get another badge.

Just making myself an issue so I don’t forget.

### PR DESCRIPTION
**Pull request by samcunliffe**
# Context/Description 

Adds a coverage build to CMakeLists and additions to the GitHub action to upload it all to codecov.io.

* Fixes  #120 

## More details

* Adds a coverage build to CMakeLists. Currently, this is only for the `tdms_tests` executable. Could also add it to `tdms` itself if we wanted to check the system tests' coverage.
* In our CI GitHub action, I install `lcov` (for nice summaries of the coverage stats for humans, and for upload to [codecov.io](https://codecov.io/gh/UCL/TDMS)).
* Add a badge 🎉 (but we need to get a token before it will be meaningful).

## Testing
Tested over on my fork.

* An example of the [build output](https://github.com/samcunliffe/TDMS/actions/runs/4796581423) (I temporarily disabled a whole bunch of irrelevant stuff in the build matrix).
* [What it looks like on codecov](https://app.codecov.io/gh/samcunliffe/TDMS/tree/coverage-build).

## Documentation 

* Added to developers' documentation with a section under "unit testing".

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,7 +4,7 @@
...
 
-# TDMS · [![latest release](https://badgen.net/github/release/UCL/TDMS)](https://github.com/UCL/TDMS/releases)  [![license](https://badgen.net/github/license/UCL/TDMS)](https://github.com/UCL/TDMS/blob/main/LICENSE) [![Build and test](https://github.com/UCL/TDMS/actions/workflows/ci.yml/badge.svg)](https://github.com/UCL/TDMS/actions/workflows/ci.yml) [![MATLAB tests](https://github.com/UCL/TDMS/actions/workflows/matlab_tests.yml/badge.svg)](https://github.com/UCL/TDMS/actions/workflows/matlab_tests.yml)
+# TDMS · [![latest release](https://badgen.net/github/release/UCL/TDMS)](https://github.com/UCL/TDMS/releases)  [![license](https://badgen.net/github/license/UCL/TDMS)](https://github.com/UCL/TDMS/blob/main/LICENSE) [![Build and test](https://github.com/UCL/TDMS/actions/workflows/ci.yml/badge.svg)](https://github.com/UCL/TDMS/actions/workflows/ci.yml) [![MATLAB tests](https://github.com/UCL/TDMS/actions/workflows/matlab_tests.yml/badge.svg)](https://github.com/UCL/TDMS/actions/workflows/matlab_tests.yml) [![codecov](https://codecov.io/gh/UCL/TDMS/branch/main/graph/badge.svg?token=FIXME)](https://codecov.io/gh/UCL/TDMS)
```
Comment (samcunliffe) [inline README.md:7]: The build has to run and we have to get an upload before I can fix this FIXME.

- [ ] Add token to the badge.

Comment (samcunliffe) [reply README.md]: Done: https://github.com/UCL/TDMS/pull/268/commits/be41c26041fb804a7bb88405d67d234167271606

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -43,7 +43,8 @@ jobs:
         # also test 'Release'.
         build_testing: [ON, OFF]
         exclude:
-          # Currently our tests don't compile in Windows so skip this.
+          # Currently our tests don't compile, and nor do we have a coverage
+          # build in Windows so skip this.
```
Comment (willGraham01) [inline .github/workflows/ci.yml:47]: :grimacing: Should this _maybe_ be an issue or are we committing to not supporting Windows?

**Review by willGraham01 — APPROVED** [review]
Only thing to flag is that we should probably make our stance on Windows-testing a bit clearer, but that's not a bug introduced with these changes.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (samcunliffe) [reply .github/workflows/ci.yml:47]: I'm inclined to say no. I softly doubt any of Peter's future developers will be using Windows.
Could tack this on to...
* #41 

...?

Comment (samcunliffe) [reply .github/workflows/ci.yml:47]: It's a completely different tool to `gcov`, though. Also, I don't know how to run it... maybe Mosè does?

Comment (giordano) [reply .github/workflows/ci.yml:47]: Sorry, what is "It" here?

Comment (samcunliffe) [reply .github/workflows/ci.yml:47]: I guess I mean [this Microsoft.CodeCoverage.Console](https://learn.microsoft.com/en-us/visualstudio/test/microsoft-code-coverage-console-tool?view=vs-2022) tool.

**Pull request merged**